File: board.py
#!/usr/bin/env python
# encoding: utf-8

import conf
import numpy as np


class Player:
    O = 1
    X = -1


class Stone(object):
    def __init__(self, pos, player):
        self.pos = pos
        self.player = player


class Board(object):
    def __init__(self):
        self.size = conf.board_size
        self.win_num = conf.win_num
        self.turn = Player.O
        self.board = np.zeros((self.size, self.size))
        self.illegal_idx = []
        self.last_pos = None

    def do_turn(self):
        if self.turn == Player.O:
            self.turn = Player.X
        else:
            self.turn = Player.O

    def get_feature(self):
        ret = np.zeros(shape=(conf.board_size, conf.board_size, 3))
        for i in range(conf.board_size):
            for j in range(conf.board_size):
                if self.board[i][j] == Player.O:
                    ret[i][j][0] = 1
                elif self.board[i][j] == Player.X:
                    ret[i][j][1] = 1
        if self.turn == Player.O:
            ret[:, :, 2] = 1
        elif self.turn == Player.X:
            ret[:, :, 2] = 0
        return ret

    def is_legal(self, stone):
        if stone.player == self.turn and self.board[stone.pos[0]][stone.pos[1]] == 0:
            return True
        else:
            return False

    def add_stone(self, stone):
        if self.board[stone.pos[0]][stone.pos[1]] != 0:
            raise RuntimeError("not empty position!")
        else:
            self.board[stone.pos[0]][stone.pos[1]] = stone.player
            self.last_pos = (stone.pos[0], stone.pos[1])

    def __str__(self):
        strs = '\n'
        for i in range(self.size):
            row = ''
            for j in range(self.size):
                if (i, j) == self.last_pos and self.board[i][j] == Player.O:
                    row += 'O)'
                elif (i, j) == self.last_pos and self.board[i][j] == Player.X:
                    row += 'X)'
                elif self.board[i][j] == Player.O:
                    row += 'O '
                elif self.board[i][j] == Player.X:
                    row += 'X '
                else:
                    row += '- '
            strs += row + '\n'
        return strs

    def count_on_direction(self, stone, xdirection, ydirection):
        i, j = stone.pos
        color = stone.player
        count = 0
        for step in range(1, 5):  # 除当前位置外,朝对应方向再看4步
            if xdirection != 0 and (j + xdirection * step < 0 or j + xdirection * step >= self.size):
                break
            if ydirection != 0 and (i + ydirection * step < 0 or i + ydirection * step >= self.size):
                break
            if self.board[i + ydirection * step][j + xdirection * step] == color:
                count += 1
            else:
                break
        return count

    def step(self, stone):
        if not self.is_legal(stone):
            raise RuntimeError("illegal idx")
        self.add_stone(stone)
        self.illegal_idx.append(stone.pos[0] * conf.board_size + stone.pos[1])
        self.do_turn()
        return self.check_done(stone)

    def check_done(self, stone):
        # 四个方向计数 竖 横 左斜 右斜
        directions = [
            [(-1, 0), (1, 0)],
            [(0, -1), (0, 1)],
            [(-1, 1), (1, -1)],
            [(-1, -1), (1, 1)],
        ]
        for axis in directions:
            axis_count = 1
            for xdirection, ydirection in axis:
                axis_count += self.count_on_direction(stone, xdirection, ydirection)
                if axis_count >= self.win_num:
                    return True, stone.player
        else:
            if len(self.illegal_idx) >= conf.board_size ** 2:
                return True, None
            else:
                return False, None

File: az/board.cpp
#include <iostream>
#include "board.h"

Board::Board(const renju::GameConf &conf) {
    this -> size = conf.board_size();
    this -> win_num = conf.win_num();
    this -> current_player = BLACK;
    for (int i = 0; i < size; i ++) {
        this -> board.emplace_back(size, 0);
    }
}

void Board::do_turn() {
    if (current_player == BLACK) {
        current_player = WHITE;
    } else {
        current_player = BLACK;
    }

}

std::vector<std::vector<std::vector<int>>> Board::get_feature() {
    return std::vector<std::vector<std::vector<int>>>();
}

bool Board::is_legal(Stone stone) {
    return stone.player == current_player and board[stone.x][stone.y] == 0;
}

void Board::add_stone(Stone stone) {
    board[stone.x][stone.y] = stone.player;
    last_pos = std::make_tuple(stone.x, stone.y);
    illegal_idx.push_back(stone.x * size + stone.y);
}

std::tuple<bool, Player> Board::check_done(Stone stone) {
    std::vector<std::vector<std::tuple<int, int>>> directions {
            {{-1, 0}, {1, 0}},
            {{0, -1}, {0, 1}},
            {{-1, 1}, {1, -1}},
            {{-1, -1}, {1, 1}}
    };
    for (auto& axis: directions) {
        int axis_count = 1;
        for (auto& xy: axis) {
            axis_count += count_on_direction(stone, std::get<0>(xy), std::get<1>(xy));
            if (axis_count >= win_num) {
                return std::make_tuple(true, stone.player);
            }
        }
    }
    return std::make_tuple(illegal_idx.size() >= size * size, NOONE);
}

std::tuple<bool, Player> Board::step(Stone stone) {
    if (!is_legal(stone)) {
        throw std::runtime_error("illegal idx");
    }
    add_stone(stone);
    do_turn();
    return check_done(stone);
}

int Board::count_on_direction(Stone stone, int xdirection, int ydirection) {
    int count = 0;
    for (int step = 1; step < 5; step ++) {
        if ((xdirection != 0) && (stone.y + xdirection * step < 0 || stone.y + xdirection * step >= size)) {
            break;
        }
        if ((ydirection != 0) && (stone.x + ydirection * step < 0 || stone.x + ydirection * step >= size)) {
            break;
        }
        if (board[stone.x + ydirection * step][stone.y + xdirection * step] == stone.player) {
            count ++;
        } else {
            break;
        }
    }
    return count;
}

void Board::print() {
    std::cout << "=======board=======" << std::endl;
    std::string str = "";
    for (int i = 0; i < size; i ++) {
        std::string row = "";
        for (int j = 0; j < size; j ++) {
            if (i == std::get<0>(last_pos) && j == std::get<1>(last_pos) && board[i][j] == BLACK) {
                row += "O)";
            } else if(i == std::get<0>(last_pos) && j == std::get<1>(last_pos) && board[i][j] == WHITE){
                row += "X)";
            } else if(board[i][j] == BLACK){
                row += "O ";
            } else if(board[i][j] == WHITE) {
                row += "X ";
            } else {
                row += "- ";
            }
        }
        str += (row + "\n");
    }
    std::cout << str;
}
